src/Microsoft.OData.Core/UriParser/Parsers/SegmentKeyHandler.cs

```diff
@@ -163,8 +163,9 @@ private static KeySegment CreateKeySegment(ODataPathSegment segment, KeySegment
                     key = KeyFinder.FindAndUseKeysFromRelatedSegment(key, keyProperties, currentNavPropSegment.NavigationProperty, previousKeySegment);
                 }
 
-                // if we still didn't find any keys, then throw an error.
-                if (keyProperties.Count != key.ValueCount && resolver.GetType() == typeof(ODataUriResolver))
+                // If we still didn't find any keys, throw an error, except for case of
+                // AlternateKeysODataUriResolver which can handle alternate keys and can throw later for unresolvable case.
+                if (keyProperties.Count != key.ValueCount && !(resolver is AlternateKeysODataUriResolver))
                 {
                     throw ExceptionUtil.CreateBadRequestError(ErrorStrings.BadRequest_KeyCountMismatch(targetEntityType.FullName()));
                 }

```

Comment:
This would break an existing custom resolver, like the ALternateKeysODataUriResolver, that handled other key resolution logic. What if we moved this check to the virtual ResolveKeys in the ODataUriParser? The default logic for that method assumes the keyvaluepairs being compared match the key values defined for the type and would have to be overridden by a derived ODataUriParser that inserted custom logic.

Reply:
yes.

Reply:
@mikepizzo : For other resolver provided by client, having to override the default ResolveKeys is a breaking change(otherwise it won't work) is also a breaking change as well. Right?

Reply:
The logic in the virtual ResolveKeys has a hard requirement that the key segments specified in the URL are 1:1 with the defined key for the type. Any custom OData URI resolver that supports key segments that are not 1:1 with the defined key must already override this method.  So putting the logic here, in the method that requires that the segments are 1:1, is safe and makes sense.